sweep/aggregator.go

```diff
@@ -742,6 +747,64 @@ func (b *BudgetAggregator) sortInputs(inputs []SweeperInput) []SweeperInput {
 	return sortedInputs
 }
 
+// splitOnLocktime splits the list of inputs based on their locktime.
+//
+// TODO(yy): this is a temporary hack as the blocks are not synced among the
+// contractcourt and the sweeper.
+func splitOnLocktime(inputs []SweeperInput) map[uint32][]SweeperInput {
+	result := make(map[uint32][]SweeperInput)
+	noLocktimeInputs := make([]SweeperInput, 0, len(inputs))
+
+	// mergeLocktime is the locktime that we use to merge all the
+	// nolocktime inputs into.
+	var mergeLocktime uint32
+
+	// Iterate all inputs and split them based on their locktimes.
+	for _, inp := range inputs {
+		locktime, required := inp.RequiredLockTime()
+		if !required {
+			log.Tracef("No locktime required for input=%v",
+				inp.OutPoint())
+
+			noLocktimeInputs = append(noLocktimeInputs, inp)
+			continue
+		}
+
+		log.Tracef("Split input=%v on locktime=%v", inp.OutPoint(),
+			locktime)
+
+		// Init the slice if not found.
+		inputList, ok := result[locktime]
```

Comment:
What about trying to merge based on lock time compatibility? If something has a lock time of `5`, then it can be bundled with a transaction that has a lock time of `3`. Or would we rather optimize for inclusion into the mempool as soon as possible?

Reply:
if the `nLockTime`s are different they cannot be bundled tho. I think most of the time this grouping logic won't be hit, as if the inputs have different locktimes would end up having different deadlines, and when we are here these inputs are already grouped by their deadlines, so they should share the same locktimes.
This is more of mitigation against this rare case brought by @morehouse [here](https://github.com/lightningnetwork/lnd/pull/8514#discussion_r1556002203) - that when two blocks arrive at the same time, it's possible we can end up mixing two inputs if their `CLTVDelta`s have a difference of 1 block, which IMO even makes this even less likely to happen as I don't think people choose CLTVDelta this way.

Reply:
>  But an attacker could easily force it to happen so they can steal funds.

How so?

Reply:
> How so?

The attacker gets to choose CLTV deltas for each hop.  LND will accept any delta in the range 80-2016 by default.

So it is trivial for the attacker to create HTLCs with the same deadlines and different timelocks:

1. At block `X`, attacker offers HTLC1 with incoming CLTV of `Y` and outgoing CLTV of `Y-82`.
2. At block `X+1`, attacker offers HTLC2 with incoming CLTV of `Y` and outgoing CLTV of `Y-81`.
3. At block `X+2`, attacker offers HTLC3 with incoming CLTV of `Y` and outgoing CLTV of `Y-80`.

Each outgoing HTLC has a different locktime, while the deadline for each to confirm is the same (`Y`)

Reply:
I think we config the incoming HTLC's cltv delta ourselves via `--bitcoin.timelockdelta=`? plus the big assumption here is we need to see two blocks come in at the same time, otherwise they'd be in separate sweeps, which I won't say it's easy.

Reply:
> I think we config the incoming HTLC's cltv delta ourselves via `--bitcoin.timelockdelta=`?

That only sets the value LND advertises.  LND will actually accept HTLCs with larger deltas:  https://github.com/lightningnetwork/lnd/blob/07b6af41dbe2a5a1c85e5c46cc41019b64640d90/htlcswitch/link.go#L2861-L2876

> plus the big assumption here is we need to see two blocks come in at the same time, otherwise they'd be in separate sweeps, which I won't say it's easy.

Yes, it's difficult to make that happen.  And this PR currently prevents all batching on restart, so the attacker can't force a bad batch by crashing LND.

Reply:
> I'm not sure whether this code allows batching in the second case or not.

I don't think this PR takes advantage of the case that the remote HTLC can be swept with a locktime greater than and can therefore be bundled with other htlcs or ?

Reply:
> That only sets the value LND advertises. LND will actually accept HTLCs with larger deltas:

Weird, as the comments say check for equality - think this is unintended?

Reply:
> I don't think this PR takes advantage of the case that the remote HTLC can be swept with a locktime greater than and can therefore be bundled with other htlcs or ?

I think they are added via `sweepMatureOutputs`.